server/datastore/mysql/scripts.go

```diff
@@ -1594,3 +1607,156 @@ func (ds *Datastore) getOrGenerateScriptContentsID(ctx context.Context, contents
+
+	// Check that all hosts exist before attempting to process them
+	for _, hostID := range hostIDs {
+		host, err := ds.Host(ctx, hostID)
+		if err != nil {
+			return "", fmt.Errorf("unable to load host information for %d: %w", hostID, err)
+		}
+
+		// All hosts must be on the same team as the script
+		sameTeamNoTeam := host.TeamID == nil && script.TeamID == nil
+		sameTeamNumber := host.TeamID != nil && script.TeamID != nil && *host.TeamID == *script.TeamID
+		sameTeam := sameTeamNoTeam || sameTeamNumber
+		if !sameTeam {
+			return "", ctxerr.Errorf(ctx, "all hosts must be on the same team as the script")
+		}
+
+		fullHosts = append(fullHosts, host)
+	}
+
+	executions := make([]fleet.BatchExecutionHost, 0, len(fullHosts))
+
+	batchExecID := ""
+
+	if err := ds.withTx(ctx, func(tx sqlx.ExtContext) error {
+		for _, host := range fullHosts {
+			// Non-orbit-enrolled host (iOS, android)
+			noNodeKey := host.OrbitNodeKey == nil || *host.OrbitNodeKey == ""
+			// Scripts disabled on host
+			scriptsDisabled := host.ScriptsEnabled != nil && !*host.ScriptsEnabled
+
+			if noNodeKey || scriptsDisabled {
+				executions = append(executions, fleet.BatchExecutionHost{
+					HostID: host.ID,
+					Error:  &fleet.BatchExecuteIncompatibleFleetd,
+				})
+				continue
+			}
+
+			if !fleet.ValidateScriptPlatform(script.Name, host.Platform) {
+				executions = append(executions, fleet.BatchExecutionHost{
+					HostID: host.ID,
+					Error:  &fleet.BatchExecuteIncompatiblePlatform,
+				})
+				continue
+			}
+
+			executionID, _, err := ds.insertNewHostScriptExecution(ctx, tx, &fleet.HostScriptRequestPayload{
+				HostID:          host.ID,
+				UserID:          userID,
+				ScriptID:        &script.ID,
+				ScriptContentID: script.ScriptContentID,
+			}, false)
+			if err != nil {
+				return ctxerr.Wrap(ctx, err, "queueing script for bulk execution")
+			}
+
+			executions = append(executions, fleet.BatchExecutionHost{
+				HostID:      host.ID,
+				ExecutionID: &executionID,
+			})
+		}
+
+		batchExecID = uuid.New().String()
+		res, err := tx.ExecContext(
+			ctx,
+			"INSERT INTO batch_script_executions (execution_id, script_id) VALUES (?, ?)",
+			batchExecID,
+			script.ID,
+		)
+		if err != nil {
+			return ctxerr.Wrap(ctx, err, "failed to insert new batch execution")
+		}
+
+		batchID, _ := res.LastInsertId()
+
+		questionMarks := strings.Join(
+			slices.Repeat([]string{"(?, ?, ?, ?)"}, len(executions)),
+			", ",
+		)
+		args := make([]any, 0, len(executions))
+		for _, execHost := range executions {
+			args = append(args, batchID, execHost.HostID, execHost.ExecutionID, execHost.Error)
+		}
+
+		insertStmt := fmt.Sprintf(`INSERT INTO batch_script_execution_host_results (batch_execution_id, host_id, host_execution_id, error) VALUES %s`, questionMarks)
+
+		if _, err := tx.ExecContext(ctx, insertStmt, args...); err != nil {
+			return ctxerr.Wrap(ctx, err, "associating script executions with batch job")
+		}
```

Comment:
Could we use [`NamedExec`](https://github.com/jmoiron/sqlx?tab=readme-ov-file#usage) here?  It lets you supply an array of maps.

Reply:
Oh I didn't know about that one, sure!

Reply:
Done deezy